### PR TITLE
feat(router): improve typings for RouterLink boolean inputs

### LINKED ISSUE #47100: feat(router): improve typings for RouterLink boolean inputs
### Which @angular/* package(s) are relevant/related to the feature request?

router

### Description

Boolean inputs of `RouterLink` (and `RouterLinkWithHref`) should also accept empty string `''` (and coerce that value to `true`).

Actually, such inputs are already coerced to boolean when they are read. Instead, they should be coerced by a setter to allow simpler template bindings, such as `<a routerLink="/" skipLocationChange>` instead of `<a routerLink="/" [skipLocationChange]="true">`. Currently, a compile-time error is thrown with the former syntax.

These inputs are: `preserveFragment`, `skipLocationChange` and `replaceUrl`.

### Proposed solution

Add getter/setter to `preserveFragment`, `skipLocationChange` and `replaceUrl` fields of `RouterLink` and `RouterLinkWithHref` directives, for example:

```ts
private _skipLocationChange: boolean = false;

get skipLocationChange(): boolean { return this._skipLocationChange; }

@Input() set skipLocationChange(skipLocationChange: boolean | '' | null | undefined) {
  this._skipLocationChange = coerceToBoolean(skipLocationChange);
}
```

### Alternatives considered

No alternative, more verbose template syntax is required.

### PR DESCRIPTION
**Pull request by EmmanuelRoux**
fixes #47100 (also contributes a bit to #24571)

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #47100 

## What is the new behavior?
See issue description

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/router/src/directives/router_link.ts [hunk 1]
```diff
@@ -120,6 +120,10 @@ import {UrlTree} from '../url_tree';
   standalone: true,
 })
 export class RouterLink implements OnChanges {
+  private _preserveFragment: boolean = false;
+  private _skipLocationChange: boolean = false;
+  private _replaceUrl: boolean = false;
```
Comment (atscott) [inline packages/router/src/directives/router_link.ts:125]: style nit: `boolean` is not necessary when you have the initializer

https://google.github.io/styleguide/tsguide.html#type-inference

**Review by atscott — APPROVED** [review]

**Review by atscott — APPROVED** [review]

**Review by atscott — APPROVED** [review]
reviewed-for: fw-router

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

File: packages/router/src/directives/router_link.ts [hunk 2]
```diff
@@ -195,6 +175,51 @@ export class RouterLink implements OnChanges {
...
+  /**
+   * Passed to {@link Router#createUrlTree Router#createUrlTree} as part of the
+   * `UrlCreationOptions`.
+   * @see {@link UrlCreationOptions#preserveFragment UrlCreationOptions#preserveFragment}
+   * @see {@link Router#createUrlTree Router#createUrlTree}
+   */
+  @Input()
+  set preserveFragment(preserveFragment: boolean | '' | null | undefined) {
+    this._preserveFragment = coerceToBoolean(preserveFragment);
+  }
```
Comment (AndrewKushnir) [inline packages/router/src/directives/router_link.ts:191]: Nit: I think it'd be great to colocate getter and setter functions (put getter underneath setter). Right now they are separated by the comment, which makes it a bit harder to read.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
This PR needs more work:
* CI is read (lint / formatting errors)
* I believe that we should have tests for the new values of the input

**Comment by EmmanuelRoux** [comment]
Widened accepted input type to `boolean|string|null|undefined` instead of `boolean|''|null|undefined`, to support both `""` and `"true"` (and be consistent with what `coerceToBoolean` accepts).

**Comment by AndrewKushnir** [comment]
@EmmanuelRoux thanks for the update. I've re-requested reviews from the current set of reviewers.

**Review by jessicajaniuk — APPROVED** [review]

**Review by pkozlowski-opensource — APPROVED** [review]

**Review by atscott — APPROVED** [review]

**Comment by pkozlowski-opensource** [comment]
This PR was merged into the repository by commit 422323cee00192239d1ab279e6a7c3a808098169.

**Pull request closed without merging**